### PR TITLE
Implement transaction verification

### PR DESCRIPTION
**Pull request by MichaelKim20**
Contains tests for transaction verification

### CODE REVIEW

File: source/agora/test/Transaction.d
Comment (AndrejMitrovic) [inline source/agora/test/Transaction.d]: The input and outputs also need to be validated so they're not negative.

Comment (MichaelKim20) [reply source/agora/test/Transaction.d]: Yes, it is. I made a mistake.

**Comment by AndrejMitrovic** [comment]
When @Geod24 mentioned that we should have code and tests together, what he meant was that they should be made at the same time and be submitted in the same pull request.

But the verification code itself should live in another module, not in a unittest one. Because the verification code will be used both in tests and within the node itself.

Does that make sense?

**Comment by MichaelKim20** [comment]
> When @Geod24 mentioned that we should have code and tests together, what he meant was that they should be made at the same time and be submitted in the same pull request.
> 
> But the verification code itself should live in another module, not in a unittest one. Because the verification code will be used both in tests and within the node itself.
> 
> Does that make sense?

Yeah, I got it.

**Comment by MichaelKim20** [comment]
Added class TransactionManager.
This class has the ability to create and store transactions.
It can also verify transactions.
When @Geod24  improves the hash function, I will use that hash function.
It will be connected to the recipient of the transaction that @AndrejMitrovic  is working on. So, I will add the transaction so that it can be verified at the receiving end of the transaction.

**Comment by AndrejMitrovic** [comment]
> It will be connected to the recipient of the transaction that @AndrejMitrovic is working on. So, I will add the transaction so that it can be verified at the receiving end of the transaction.

I think this part can be added in a separate pull request later, so your work is not blocked on my work.

So the only blocker right now for both of us is hashing.

**Comment by MichaelKim20** [comment]
> > It will be connected to the recipient of the transaction that @AndrejMitrovic is working on. So, I will add the transaction so that it can be verified at the receiving end of the transaction.
> 
> I think this part can be added in a separate pull request later, so your work is not blocked on my work.
> 
> So the only blocker right now for both of us is hashing.

We need to find 'Output' in 'Block' using 'Input' ('previous' and 'index').
This should also be done later after the two PR's are Merge. What do you think?

File: source/agora/common/TransactionManager.d
Comment (Geod24) [inline source/agora/common/TransactionManager.d]: This does not use any of the class' variable. So it could be made `static` to allow it to be called without an instance. Or it could be made a free function.

However, I think it can be removed. Because you can use:
```d
auto txout = TxOut(address, value);
```

the same way you would use:
```d
auto txout = txmanager.newTxOut(address, value);
```

D generate a default constructor for `struct` which is equivalent to:
```d
this(PublicKey _1 = PublicKey.init, Amount _2 = Amount.init)
```

So you can call `TxOut`, `TxOut(address)` and `TxOut(address, amount)`.

File: source/agora/common/TransactionManager.d
Comment (Geod24) [inline source/agora/common/TransactionManager.d]: Same comment as `newTxOut`, this can be removed

File: source/agora/common/TransactionManager.d
Comment (Geod24) [inline source/agora/common/TransactionManager.d]: We should not serialize to JSON here, the struct should be directly hashed. See the hashing primitive PR for an example on how to do this.

File: source/agora/common/TransactionManager.d
Comment (Geod24) [inline source/agora/common/TransactionManager.d]: There shouldn't be any dependency to Vibe.d here

File: source/agora/common/TransactionManager.d
Comment (Geod24) [inline source/agora/common/TransactionManager.d]: I see that this is never initialized. So it always have the default value, which can be represented as `Hash.init`.
It is better to use `Hash.init` instead of `nullTxHash`, as `Hash.init` can be optimized (https://en.wikipedia.org/wiki/Constant_folding) while `nullTxHash` cannot.

File: source/agora/common/TransactionManager.d
Comment (Geod24) [inline source/agora/common/TransactionManager.d]: Same comment as `nullTxHash`: Use `Signature.init`

File: source/agora/common/TransactionManager.d
Comment (Geod24) [inline source/agora/common/TransactionManager.d]: You can simplify this with:
```d
return tx.inputs.length == 1 && tx.inputs[0] == TxIn.init;
```

File: source/agora/common/TransactionManager.d
Comment (Geod24) [inline source/agora/common/TransactionManager.d]: Style nitpick:
```suggestion
    public Output* findTxOutput (Hash hash, uint index)
```

File: source/agora/common/TransactionManager.d
Comment (Geod24) [inline source/agora/common/TransactionManager.d]: This is doing a double lookup. The builtin hashmap returns a pointer to the element if it exists, or `null`.
Additionally, you can define a new variable in `if` statement, so the following code:
```d
if (auto tx = hash in this.txs)
{
    // Use `tx`
}
else
{
    // Cannot use `tx` from here, but we know it is `null`
    return null;
}
```

Works just fine.

File: source/agora/common/TransactionManager.d
Comment (Geod24) [inline source/agora/common/TransactionManager.d]: I think you should take a look at [`std.algorithm`](https://dlang.org/phobos/std_algorithm.html).
Many of those loop could be turned into calls to `each`, `map` & co.

**Review by Geod24 — COMMENTED** [review]
First round of review done

**Comment by MichaelKim20** [comment]
> First round of review done

@Geod24 All the requested modifications have been completed.

**Comment by Geod24** [comment]
Great work !

One point though, if a function does not use any member of a class, and its usage is not specific to a class, it should not be part of a class. Other parts of the program might need to get the `sum` of the input. They should not have to have an instance of a `TransactionManager` to do so.

For example, imagine I want to print the sum of all transactions in a debug logging statement in the network code.

Using https://github.com/bpfkorea/agora/blob/740d54a677a209dabcb8fbd1893c8f2950089b37/source/agora/node/Network.d#L300-L306 as a base

```d
    public void sendMessage (Transaction tx) @safe
    {
        scope tm = new TransactionManager();
        foreach (ref node; this.peers)
        {
            logDebug("Sending a transaction with %d input and %d output", tm.getInputSum(tx), tm.getOutputSum(tx));
            node.sendMessage(tx);
        }
    }
```

Instead, if you make it a free function:
```d
public long getOutputSum (Transaction tx)
{
    return tx.outputs.map!(a => a.value).sum();
}
```

I can call it like this:
```d
    public void sendMessage (Transaction tx) @safe
    {
        foreach (ref node; this.peers)
        {
            logDebug("Sending a transaction with %d input and %d output", tx.getInputSum(), tx.getOutputSum());
            node.sendMessage(tx);
        }
    }
```

Which I think is much clearer. It does not have to be a member of `Transaction` because we can use UFCS, as you already did with `map` and `sum`.

File: source/agora/common/TransactionManager.d
Comment (AndrejMitrovic) [inline source/agora/common/TransactionManager.d]:
```suggestion

import std.algorithm;
```

File: source/agora/common/TransactionManager.d
Comment (AndrejMitrovic) [inline source/agora/common/TransactionManager.d]: Do you plan on ever overriding and changing what this method does? Because if not, we don't really need it. Unittest code can simply call `hashFull(tx)`.

File: source/agora/common/TransactionManager.d
Comment (AndrejMitrovic) [inline source/agora/common/TransactionManager.d]:
```suggestion
                .filter!(a => a !is null)
```

@Geod24: wasn't there some warning / deprecation in the past about using `!=` with null?

**Comment by AndrejMitrovic** [comment]
I agree with @Geod24's comments in https://github.com/bpfkorea/agora/pull/60#issuecomment-508912767.

File: source/agora/common/TransactionManager.d
Comment (Geod24) [inline source/agora/common/TransactionManager.d]:
```suggestion

import agora.common.Data;
```

File: source/agora/common/TransactionManager.d
Comment (Geod24) [inline source/agora/common/TransactionManager.d]:
```suggestion
    Class for the transactions verification
```

File: source/agora/common/TransactionManager.d
Comment (Geod24) [inline source/agora/common/TransactionManager.d]:
```suggestion
    *******************************************************************************/
    
```

File: source/agora/common/TransactionManager.d
Comment (MukeunKim) [reply source/agora/common/TransactionManager.d]: > Great work !
> 
> One point though, if a function does not use any member of a class, and its usage is not specific to a class, it should not be part of a class. Other parts of the program might need to get the `sum` of the input. They should not have to have an instance of a `TransactionManager` to do so.
> 
> For example, imagine I want to print the sum of all transactions in a debug logging statement in the network code.
> 
> Using
> 
> https://github.com/bpfkorea/agora/blob/740d54a677a209dabcb8fbd1893c8f2950089b37/source/agora/node/Network.d#L300-L306
> 
> as a base
> ```d
>     public void sendMessage (Transaction tx) @safe
>     {
>         scope tm = new TransactionManager();
>         foreach (ref node; this.peers)
>         {
>             logDebug("Sending a transaction with %d input and %d output", tm.getInputSum(tx), tm.getOutputSum(tx));
>             node.sendMessage(tx);
>         }
>     }
> ```
> 
> Instead, if you make it a free function:
> 
> ```d
> public long getOutputSum (Transaction tx)
> {
>     return tx.outputs.map!(a => a.value).sum();
> }
> ```
> 
> I can call it like this:
> 
> ```d
>     public void sendMessage (Transaction tx) @safe
>     {
>         foreach (ref node; this.peers)
>         {
>             logDebug("Sending a transaction with %d input and %d output", tx.getInputSum(), tx.getOutputSum());
>             node.sendMessage(tx);
>         }
>     }
> ```
> 
> Which I think is much clearer. It does not have to be a member of `Transaction` because we can use UFCS, as you already did with `map` and `sum`.

The function is dependent on the internal data structure.
Please check in detail to see if you can do this.
Wouldn't it be better to make improvements after the blocks are made in the future?

Comment (MichaelKim20) [reply source/agora/common/TransactionManager.d]: Modified~

File: source/agora/common/Transaction.d
Comment (Geod24) [inline source/agora/common/Transaction.d:25]:
```suggestion

import std.algorithm;
```

File: source/agora/common/Transaction.d
Comment (Geod24) [inline source/agora/common/Transaction.d]:
```suggestion
public bool isCoinbaseTx (Transaction tx)
```

File: source/agora/common/Transaction.d
Comment (Geod24) [inline source/agora/common/Transaction.d]:
```suggestion
public Transaction newCoinbaseTX (PublicKey address, Amount value = 0)
```

File: source/agora/common/Transaction.d
Comment (Geod24) [inline source/agora/common/Transaction.d]:
```suggestion

public Amount getSumInput (Transaction tx, Output* delegate(Hash hash, size_t index) findOutput)
```

Changes:
- `public`
- No space after `delegate` keyword
- Indexes are always `size_t`

File: source/agora/common/Transaction.d
Comment (Geod24) [inline source/agora/common/Transaction.d]:
```suggestion
public long getSumOutput (Transaction tx)
```

File: source/agora/common/Transaction.d
Comment (Geod24) [inline source/agora/common/Transaction.d]:
```suggestion
public bool verifyData (Transaction tx, Output* delegate (Hash hash, size_t index) findOutput)
```

File: source/agora/common/Transaction.d
Comment (Geod24) [inline source/agora/common/Transaction.d]: Move this `if` before `if (tx.isCoinbaseTx())` ?

File: source/agora/test/Transaction.d
Comment (Geod24) [inline source/agora/test/Transaction.d]: Those unittests don't use the network. They are good though, and should be moved to the `TransactionManager` code. They are "simple" `unittest` as opposed to ["Network unittests"](https://github.com/bpfkorea/agora/pull/74/files#diff-f69cd4b5588d09bb541f85596fdacad3R193).

**Review by Geod24 — COMMENTED** [review]
Pretty good, a few comments, mostly style

File: source/agora/common/TransactionManager.d
Comment (MichaelKim20) [reply source/agora/common/TransactionManager.d]: This file is deleted, moved to source/agora/common/Transaction.d

File: source/agora/common/TransactionManager.d
Comment (MichaelKim20) [reply source/agora/common/TransactionManager.d]: This file is deleted, moved to source/agora/common/Transaction.d

File: source/agora/common/TransactionManager.d
Comment (MichaelKim20) [reply source/agora/common/TransactionManager.d]: This file is deleted, moved to source/agora/common/Transaction.d

File: source/agora/common/Transaction.d
Comment (MichaelKim20) [reply source/agora/common/Transaction.d]: Modified

File: source/agora/test/Transaction.d
Comment (MichaelKim20) [reply source/agora/test/Transaction.d]: Modified

File: source/agora/common/Transaction.d
Comment (MichaelKim20) [reply source/agora/common/Transaction.d]: Modified

**Review by Geod24 — APPROVED** [review]

**Comment by MichaelKim20** [comment]
@Geod24 @AndrejMitrovic
I have organized it into a single file(source/agora/common/Transaction.d).

**Pull request merged**
